src/parcsr_ls/par_relax.c

```diff
@@ -14,4397 +14,1398 @@
+            ii = A_diag_j[jj];
+            res -= A_diag_data[jj] * Vtemp_data[ii];
          }
-         if (num_procs > 1)
+         for (jj = A_offd_i[i]; jj < A_offd_i[i+1]; jj++)
          {
-            hypre_ParCSRCommHandleDestroy(comm_handle);
-            comm_handle = NULL;
+            ii = A_offd_j[jj];
+            res -= A_offd_data[jj] * v_ext_data[ii];
          }
 
-         /*-----------------------------------------------------------------
-          * Relax all points.
-          *-----------------------------------------------------------------*/
-
-         if (relax_points == 0)
+         if (Skip_diag)
          {
-#ifdef HYPRE_USING_OPENMP
-#pragma omp parallel for private(i,ii,jj,res) HYPRE_SMP_SCHEDULE
+            u_data[i] *= one_minus_weight;
+            u_data[i] += relax_weight * res / di;
+         }
+         else
+         {
+            u_data[i] += relax_weight * res / di;
+         }
+      }
+   }
+
+   if (num_procs > 1)
+   {
+      hypre_TFree(v_ext_data, HYPRE_MEMORY_HOST);
+      hypre_TFree(v_buf_data, HYPRE_MEMORY_HOST);
+   }
+
+   return hypre_error_flag;
+}
+
+HYPRE_Int
+hypre_BoomerAMGRelax0WeightedJacobi( hypre_ParCSRMatrix *A,
+                                     hypre_ParVector    *f,
+                                     HYPRE_Int          *cf_marker,
+                                     HYPRE_Int           relax_points,
+                                     HYPRE_Real          relax_weight,
+                                     hypre_ParVector    *u,
+                                     hypre_ParVector    *Vtemp )
+{
+   return hypre_BoomerAMGRelaxWeightedJacobi_core(A, f, cf_marker, relax_points, relax_weight, NULL, u, Vtemp, 1);
+}
+
+HYPRE_Int
+hypre_BoomerAMGRelax18WeightedL1Jacobi( hypre_ParCSRMatrix *A,
+                                        hypre_ParVector    *f,
+                                        HYPRE_Int          *cf_marker,
+                                        HYPRE_Int           relax_points,
+                                        HYPRE_Real          relax_weight,
+                                        HYPRE_Real         *l1_norms,
+                                        hypre_ParVector    *u,
+                                        hypre_ParVector    *Vtemp )
+{
+#if defined(HYPRE_USING_CUDA)
+   //HYPRE_ExecutionPolicy exec = hypre_GetExecPolicy2( hypre_VectorMemoryLocation(x), hypre_VectorMemoryLocation(b) );
+   //RL: TODO back to hypre_GetExecPolicy2 later
+   HYPRE_ExecutionPolicy exec = HYPRE_EXEC_DEVICE;
+   // TODO implement CF relax on GPUs
+   if (relax_points != 0)
+   {
+      exec = HYPRE_EXEC_HOST;
+   }
+
+   if (exec == HYPRE_EXEC_DEVICE)
+   {
+      // XXX GPU calls Relax7 XXX
+      return hypre_BoomerAMGRelax7Jacobi(A, f, relax_points, relax_weight, l1_norms, u, Vtemp);
+   }
+   else
 #endif
-            for (i = 0; i < n; i++)
-            {
+   {
+      /* in the case of non-CF, use relax-7 which is faster */
```

Comment:
Are you sure about this? In my experience 0 is faster than 7 on the CPU.

Reply:
@ulrikeyang This is to enforce consistency with the old code. Non-CF `relax-type == 18` used to call Jacobi in `hypre_ParCSRRelax` in `ams.c` that uses Matvec, which is the same as `relax-type == 7`. Also, 0 doesn't use L1-norm.